nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/PutDatabaseRecord.java

```diff
@@ -0,0 +1,1058 @@
+
+            // Get the statement type from the attribute if necessary
+            String statementType = statementTypeProperty;
+            if (USE_ATTR_TYPE.equals(statementTypeProperty)) {
+                statementType = flowFile.getAttribute(STATEMENT_TYPE_ATTRIBUTE);
+            }
+            if (StringUtils.isEmpty(statementType)) {
+                log.error("Statement Type is not specified, flowfile {} will be penalized and routed to failure", new Object[]{flowFile});
+                flowFile = session.penalize(flowFile);
+                session.transfer(flowFile, REL_FAILURE);
+            } else {
+                RecordSchema recordSchema;
+                try (final InputStream in = session.read(flowFile)) {
+
+                    final RecordReader recordParser = recordParserFactory.createRecordReader(flowFile, in, log);
+                    recordSchema = recordParser.getSchema();
+
+                    if (SQL_TYPE.equalsIgnoreCase(statementType)) {
+
+                        // Find which field has the SQL statement in it
+                        final String sqlField = context.getProperty(FIELD_CONTAINING_SQL).evaluateAttributeExpressions(flowFile).getValue();
+                        if (StringUtils.isEmpty(sqlField)) {
+                            log.error("SQL specified as Statement Type but no Field Containing SQL was found, flowfile {} will be penalized and routed to failure", new Object[]{flowFile});
+                            flowFile = session.penalize(flowFile);
+                            session.transfer(flowFile, REL_FAILURE);
+                        } else {
+                            boolean schemaHasSqlField = recordSchema.getFields().stream().anyMatch((field) -> sqlField.equals(field.getFieldName()));
+                            if (schemaHasSqlField) {
+                                try (Statement s = con.createStatement()) {
+
+                                    try {
+                                        s.setQueryTimeout(queryTimeout); // timeout in seconds
+                                    } catch (SQLException se) {
+                                        // If the driver doesn't support query timeout, then assume it is "infinite". Allow a timeout of zero only
+                                        if (queryTimeout > 0) {
+                                            throw se;
+                                        }
+                                    }
+
+                                    Record currentRecord;
+                                    while ((currentRecord = recordParser.nextRecord()) != null) {
+                                        Object sql = currentRecord.getValue(sqlField);
+                                        if (sql != null && !StringUtils.isEmpty((String) sql)) {
+                                            // Execute the statement as-is
+                                            s.execute((String) sql);
```

Comment:
In an end-to-end CDC flow, we would use CaptureChangeMySQL to get database events to stream into this PutDatabaseRecord as your template shared on Gist does. Those database events include SQL statements such as `begin` and `commit`, and such statements are executed literary here.

I think it would be great if users can configure CaptureChangeMySQL whether to emit `begin` and `commit` event or not, as these statements do not have significant meaning for synchronizing change since CaptureChangeMySQL emits FlowFile per updated record, but just for EnforceOrder to order events correctly.

If we can eliminate these events, we can minimize the number of FlowFiles which would lead us to a better performance.

Also, (for MySQL at least) `begin` and `commit` is not a database or table specific event. When I replicate changes from table A to table B using CaptureChangeMySQL and PutDatabaseRecord, I saw following behavior and felt `begin` and `commit` are a little bit disturbing:

1. Insert a row into table A
1. 3 events are emitted, begin, insert A and commit. Then CaptureChangeMySQL emitted 3 FlowFiles.
1. PutDatabaseRecord execute 3 SQLs. begin, insert B and commit.
1. Another 3 events are emitted derived from `insert B` via MySQL bin-log, begin, insert B and commit.
1. Since I have configured CaptureChangeMySQL only listens to `table A`, it only emitted 2 FlowFiles, begin and commit.
1. PutDatabaseRecord executes the begin and commit SQL.

Do we have to take these "begin" and "commit" statements into account for a CDC flow?